[WIP] Adding Gated Convolutional Network

francescopapaleo opened the pull request:
Adding a base class for the GCN network with an example on how to wrap a pre-trained model

-[x] adding a conv1dcausal class to the SDK
-[x] adding gcn1d class
-[x] adding wrapping example for a pre-trained model
-[x] black formatter

francescopapaleo commented:
@bogdanteleaga and @christhetree : what do you think about this?

bogdanteleaga commented:
Thank you Francesco!

I have a couple of thoughts on this:
- The PaddingCached and Conv1dCached modules that you have in your example file look identical to the ones we currently have in `examples/tcn_1d.py`. We don't want them duplicated and it'd be better to place them inside the `neutone_sdk` module like the TODO comments say (perhaps inside this new conv1dcausal file you're creating, maybe rename it to conv1d?). Seems like they only work for mono inputs so they should be generalized as well. For this PR you could just import them from that file and leave this for another time.
- Besides that, probably out of scope here, but I think we should move the tcn_1d, gcn_1d, conv1d files inside another module like `neutone_sdk.blocks` or `neutone_sdk.nn` or something similar.

- On a general note, I'd rather not add examples that people cannot run without an existing model file. RAVE is a bit of an exception since it's a popular model and it is helpful for users. But otherwise our other examples are standalone, so overall I think it'd be better if you keep the examples/example_spring_reverb.py file inside the parent repo (https://github.com/francescopapaleo/neutone-spring-reverb) similar to these ones:
    - https://github.com/mcomunita/neurafuzz/blob/master/src/gcntfilm_neutone_fuzzring.py
    - https://github.com/hyakuchiki/realtimeDDSP/blob/master/export.py
 
    And then we place links to them in the README for people to refer to.

- Instead we could have the example simply wrap a randomly initialized GCN similar to the TCN example.

francescopapaleo commented:
- [x] removed the example, a full example will be provided by the Colab notebook and could be linked in the README.md.
- [x] Gathered all the elements of the GCN architecture in a single file and added the TFiLM layer, so now TCN and GCN are independent.

christhetree commented on neutone_sdk/gcn_1d.py:
```diff
@@ -0,0 +1,381 @@
+
+
+class TFiLM(nn.Module):
+    def __init__(self, n_channels, cond_dim, tfilm_block_size, rnn_type="lstm"):
```
Please add type hints to all method arguments for consistency in the SDK.

francescopapaleo replied:
Done, sorry about the oversight.

christhetree replied:
Thanks, but I meant type hinting the parameters and return type for all methods, not just TFiLM, so there's quite a few still missing

christhetree commented on neutone_sdk/gcn_1d.py:
```diff
@@ -0,0 +1,381 @@
+        return rf
+
+
+if __name__ == "__main__":
```
As we mentioned in the meeting you can remove this once you're done testing

francescopapaleo replied:
Sure!

bogdanteleaga commented:
LGTM, can you add a link in the README that points to your notebook for the GCN?

francescopapaleo commented:
Type hints should be ok now and I've added a link to the demo notebook in the README.md file. 
Thank you for your help!

christhetree commented:
LGTM feel free to squash and merge @francescopapaleo
